A32-xds-circuit-breaking.md

```diff
@@ -80,13 +86,21 @@ fail the call with status UNAVAILABLE. Such failed calls will not be retried,
 but they will be recorded to the `total_dropped_requests` counts in 
 cluster-level load reports and reported to the load reporting server.
 
+Note an EDS LB policy may be responsible for handling `service_name` switch
```

Comment:
I think this may be difficult to implement, and I'm not sure it's actually necessary.  If the EDS service name is changing, it's also likely that the set of backends is changing, in which case it would be fine for the active requests to be counted separately for each set of backends.

I think we should just remove this.  If it actually becomes a problem (which seems very unlikely), we can address it later.

Reply:
Isn't our current EDS policy implementation already handling the EDS service name change? The stats are recorded separately for each set of backends (remember each `ClusterStats` are for stats per {cluster, eds_service}). But requests sent both EDS services should be aggregated together for circuit breaking limits, as logically they are requests sent to the same cluster. 

There is not really anything special to be added for the implementation, just highlighting that in-flight requests should be aggregated to the same counter for enforcing `max_request` while drops are recorded separately if they sent to different EDS services.

Reply:
Recording the drop stats is not difficult to implement at all; as you said, we're already doing that for the EDS-based drop stats.  However, having a common counter for enforcing the max-requests limit may be tricky to implement.

In C-core, we completely swap out the EDS policy instance when the EDS service name changes, and we use fairly generic logic for switching between policies.  To have a common counter between them, we would need to have special logic specific to the EDS policy to reach into the old policy instance to get the counter and then pass it to the new policy instance that we're instantiating.  This is doable, but it does add complexity.

Also, there are some semantic questions here.  What happens if a single CDS update both changes the EDS service name and updates the max-requests limit at the same time?  In that case, should we apply the new limit only to the new EDS policy instance, not the old one, or should it immediately take effect for both?

And as I said above, it's not clear to me that it's really beneficial to share the counter between the two policies.  If we're changing the EDS policy name, we're probably also changing the set of backends, so applying different counters to the two instances seems fine anyway.

@htuch I'm curious as to what Envoy does in this case.  Does envoy track the max-requests circuit breaker per cluster or per EDS service name?  I wouldn't be surprised if it tracks it per cluster, given that the circuit breaker config is part of the CDS response and not the EDS response.  If so, how does Envoy handle the edge cases I mentioned above?

Reply:
> In C-core, we completely swap out the EDS policy instance when the EDS service name changes, and we use fairly generic logic for switching between policies.

Oh I see. We are having somewhat different implementations so that our definition for EDS policy is actually different. Here EDS policy, specifically I am really referring to the class that implements the `LoadBalancer` interface. In C-core a EDS policy's scope is bounded to a {cluster:eds_service} and switching eds_service happens outside the EDS policy (it's actually inside its CDS policy). However, in Java a EDS policy is somewhat "an extension of the CDS policy", as its scope is bounded to a cluster (same as the CDS policy) and internally it maintains a switchable subcomponent that takes care of the EDS service name switch. So our "EDS policy" is really just a wrapper boxing the real EDS policy and  instantiated by its parent CDS policy, it is not doing any load balancing work except receiving the EDS config. After that, the received config is delegated to the subcomponent.

Such an implementation is actually quite awkward and unclean. For example the wrapper class does nothing and has no real semantic meaning. It contains the work delegated by the CDS policy (aka, maintain two child real EDS policy instances during EDS service name switch). Also, this only happens on CDS and EDS polices (I believe part of the reason may be they are the two that evolved from the very early version of xDS design), each layer of the downstream policies below EDS policy maintains its child policies right inside its own implementation. I believe this implementation really needs to be refactored. After that, Java's definition of "EDS policy" should be clearer. 
/cc @dapengzhang0 who implemented the existing CDS/EDS policies.

> What happens if a single CDS update both changes the EDS service name and updates the max-requests limit at the same time? In that case, should we apply the new limit only to the new EDS policy instance, not the old one, or should it immediately take effect for both?

While waiting for @htuch's answer, my understanding is that the max-requests limit should be enforcing requests sent to hosts in a (virtual) cluster. If a cluster is changing its backing service (aka, EDS service), unfinished RPCs to the old service are still within the cluster. So they should definitely contribute to the number of in-flight requests to the cluster. Since we are doing graceful switch for EDS service change, the switching process is really first expand the cluster to two services and then drop the old one. That's why I think we should use a common counter for EDS policies having the same parent CDS policy. This also implies that when a single CDS update changes both the EDS service name and the max-requests limit, it should immediately take effect for both services.

In terms of having a common counter for enforcing the max-requests limit for EDS policies with the same parent CDS policy. It is tricky because we don't have a convenient way of passing arguments from a parent policy to its child policy. We use the Update method in `LoadBalancer` interface to do so (though this is another awkward thing in the design). It is unfortunate. We would need a slight change for existing description of implementing circuit breaking: we should maintain the counter in the CDS policy while passing it as an address attribute to its child EDS policy instances, so requests sent through EDS policies with the same parent CDS policy are aggregated to the same counter.

Reply:
Envoy tracks per-cluster. EDS service name is really just a convenience for allowing clusters to modify the name they use for EDS lookup. These aren't really first class concepts inside Envoy beyond this convenient realiasing.

Reply:
It's worth pointing out that circuit breakers are not an EDS concept; they apply to any cluster type that Envoy supports.

Reply:
> It's worth pointing out that circuit breakers are not an EDS concept; they apply to any cluster type that Envoy supports.

That's a really good point.  We do plan to support at least LOGICAL_DNS and aggregate clusters in the future, and we might want to add other cluster types as well.  And in those cases, we would not be using the EDS LB policy.

Harvey, do circuit breakers apply to aggregate clusters as well?  How does that work if the underlying clusters also have their own circuit breakers defined?

Reply:
We want requests sent through the same CDS LB policy to be aggregated together and the total number is limited by the cluster-level circuit breakers. But on the other hand, among those requests dropped by circuit breakers should be recorded and reported separately for each {cluster, eds_service}, which is the scope of an EDS LB policy. Ideally we would implement this max-requests limit in the single Picker in CDS LB policy. But because we have to record drops separately for each eds_service, we need to split this work to each EDS LB policy's Picker while still aggregating to the same per-cluster counter. The aggregation is effectively equivalent as it is implemented in CDS LB policy's Picker, and conceptually it is tracking requests sent to the cluster. So I think the solution would be maintain the counter in the CDS LB policy and pass it down to the child EDS LB policies for enforcing the limit. With this approach, each EDS policy knows which requests are dropped by the circuit breaker while requests are still tracked per-cluster. @markdroth Does this sound good?

Reply:
I want to fully understand how Envoy does this before we decide how to structure our implementation.  I don't want to go down a path that results in dramatically different semantics from Envoy.

> Envoy tracks per-cluster. EDS service name is really just a convenience for allowing clusters to modify the name they use for EDS lookup. These aren't really first class concepts inside Envoy beyond this convenient realiasing.

Is that also true for load reporting?  How does load reporting work for the case where the client receives a CDS update that changes the EDS service name?  Note that each load report in an LRS request contains both the cluster name and the EDS service name, which implies that both fields are used as the "key" for the corresponding load report data.  So I'd like to understand how Envoy splits up the load reporting information when the EDS service name changes.

For example, consider the following timeline (assumes 10s load reporting interval):

- t=0: Envoy sends an LRS request including a load report for {cluster_name=C, EDS_service_name=E1}.
- t=0 through t=3: Envoy sends 10 requests to the cluster and drops 2 requests.
- t=3: Envoy receives a CDS update changing the EDS service name from E1 to E2.  Envoy requests E2 via EDS.  (Presumably it continues to use the backends from E1 until it gets new backends from E2?)
- t=4: Envoy has now gotten the EDS resource E2 and connected to those backends, so it now switches over and stops using the original backends from E1.
- t=4 through t=10: Envoy sends 15 requests to the cluster and drops 4 requests.
- t=10: Envoy sends a new LRS request.

The question is, what is included in the LRS request sent at time t=10?  Note that there are two sets of stats here: those from before the EDS service name changed and those after the EDS service name changed.  If Envoy just sends one load report for the cluster as a whole and reports the most recent EDS service name, then it will send both sets of stats as a single entry, like this:

- {cluster_name=C, EDS_service_name=E2}: requests=25, drops=6

But if it records each set of stats under its own EDS service name, then it would send this:

- {cluster_name=C, EDS_service_name=E1}: requests=10, drops=2
- {cluster_name=C, EDS_service_name=E2}: requests=15, drops=4

gRPC does the latter.  This seems more accurate, because if E1 and E2 represent disjoint sets of endpoints, then it does not really make sense to lump their stats together.  It also seems to violate the LRS semantics of using both the CDS and EDS names as the key.

By that same logic, it seems weird to track the max-requests counter at the CDS level, because the resulting stats would need to be tracked and reported separately for the two EDS service names.  This also implies that if we get a CDS update that both changes the max-requests limit and the EDS service name at the same time, the new limit should apply only to the new EDS service name.  (Although admittedly, that last part would be much clearer if the max-requests limit was set in the EDS response instead of the CDS response.)

I guess what I'm pointing out here is that it seems very inconsistent to say that the EDS resource name is not a first-class concept but that it is used as part of the key for the LRS data.  I think we need to clarify and document how this is supposed to work.

Reply:
Thanks @markdroth for posting out such a detailed example, this accurately questions what has been bothering me for a long time and I think it is very valuable for us to completely understand the whole logic in Envoy. @htuch Could you please help us figure this out? Thanks.

Reply:
For aggregate clusters, I believe we have just a single circuit breaker. There are no tests for this, I've opened up https://github.com/envoyproxy/envoy/issues/12739. I think it would also make sense to add a PR to clarify in the API.

Reply:
The EDS service name and LRS interaction is limited to https://github.com/envoyproxy/envoy/blob/cbdd01c22b84ac5333f95f030821dfcf6db49734/source/common/upstream/load_stats_reporter.cc#L77. 

When you change EDS service name, you're basically replacing the cluster. At that point, you are resetting all the stats.

Reply:
@htuch Thanks for your quick response!

> When you change EDS service name, you're basically replacing the cluster.

This makes sense to me. But I believe all the subtleties come from dealing with streaming RPCs and graceful switches for EDS service name change. When these two are involved, it's hard to make a clean replacement. 

> At that point, you are resetting all the stats.

Does this mean some stats are just thrown away when the transition happens? The situation is especially hard to make a clear cut when there are streaming RPCs ongoing to hosts in the old EDS service.

How does Envoy deal with the transition? For example, there are requests in-flight to hosts in the old EDS service and they it receives an CDS response to switch to another EDS service. Does it let the in-flight requests finish? What if the requests are streaming and never finish? Before connections to endpoints in the new EDS service are established, are new requests just get buffered instead of being sent to endpoints in the old EDS service? 

I think this would be very important to us, as our current implementation is trying to handle switchovers as mild as possible (e.g., trying to use the old EDS service before the new one becomes ready, trying to record/report every single load, etc). The effort trying to gracefully deal with the very detailed aspects of transitions makes our design somewhat diverge from what Envoy does, and some usages may not be intended in Envoy's design.

Reply:
I don't understand what "graceful switches for EDS service name change" means here. When a cluster is replaced, the new cluster first warms, and then when it completes warming, replaces the existing cluster. The older cluster connections are then drained. @snowp can most likely add some additional detail here (or correct any inaccuracy).

Reply:
The new cluster definitely replaces the old one completely after the new one is warmed, I'm not sure exactly what happens with the old one. It's possible that we just tear down the old connections without warning. 

In any case there is no handing over of connections etc., the new and old cluster are independent.

Reply:
> I don't understand what "graceful switches for EDS service name change" means here.

It's about when receiving an CDS response that contains a different EDS service name inside. Graceful switch means we will keep sending requests to hosts resolved by the old EDS service name until connection to hosts resolved by the new one is established. So this would be the same as what you are saying about warming the new cluster.

The problem is, "the older cluster connections are then drained", does it just forcefully terminate requests even if they are streaming? And I think this is something that is unclear in our design, although it is a very detailed thing. The time a streaming RPC will finish is undefined, if we let it keep going, we will wait for an unbounded amount of time to cut off the old cluster. That would keep the stats object for the old cluster for an unbounded amount of time as well. So does Envoy just throws away the unreported stats for the old cluster when it is cut off?

Reply:
AFAICT, Envoy transitions the connections to hosts in the respective connection pools on the no longer active cluster to drained state, but I don't see where GOAWAY is sent or any timeouts in the connection pool draining. This would imply they no longer contribute to LRS stats (or cluster stats), potentially indefinitely. @snowp @mattklein123 does this seem right?

Reply:
The client doesn't send GOAWAY to upstream since we don't support push currently and it wouldn't do anything. Basically we just wait until there are no active streams per other mechanisms (they finish, timeout, etc.)

Reply:
In theory we could have a long running stream, e.g. several hours, where there is an observability gap as a result of this if I read ^^ right.

Reply:
Yes, that case could be an issue. Presumably there is some idle timeout still in play in which it would drain eventually, but if we want active draining of streams a a result of the update I think that functionality would need to be added.

Reply:
> The question is, what is included in the LRS request sent at time t=10? Note that there are two sets of stats here: those from before the EDS service name changed and those after the EDS service name changed. If Envoy just sends one load report for the cluster as a whole and reports the most recent EDS service name, then it will send both sets of stats as a single entry, like this:
> 
> {cluster_name=C, EDS_service_name=E2}: requests=25, drops=6
> But if it records each set of stats under its own EDS service name, then it would send this:
> 
> {cluster_name=C, EDS_service_name=E1}: requests=10, drops=2
> {cluster_name=C, EDS_service_name=E2}: requests=15, drops=4

@markdroth Based on what has been discussed above, how exactly stats are recorded/reported for requests sending to the old EDS service during the transition seems to be a bit vague, even in Enovy it might not have a definitive answer. I can imagine the subtleties. That amount of stats during the transition may not be that important and critical. But the answer to your question above (at time t=10, the stats being reported) seems to be either one of the following two:

- {cluster_name=C, EDS_service_name=E2}: requests=15, drops=4

or 

- {cluster_name=C, EDS_service_name=E1}: requests=10, drops=2 
   {cluster_name=C, EDS_service_name=E2}: requests=15, drops=4

The former is we do not care the stats remain unreported for the old EDS service and the latter is we try to report every single load. Note for the latter report, the LRS server will combine IMO, the former is cleaner and easier to implement as the time a streaming RPC ends is unbounded.

Based on what is mentioned in https://github.com/grpc/proposal/pull/200#discussion_r473400202:
> When you change EDS service name, you're basically replacing the cluster (which is also true for other field mutations in the cluster). At that point, you are resetting all the stats.

Then it makes more sense that we use separate max-request counters for EDS policy (or reset the counter when changing to use a different EDS service name).

Reply:
I think we should fix Envoy's behavior to report both active and draining clusters, but this is more complicated than just thinking about EDS service name. There are other cluster updates that are possible that will result in similar behavior, but with no EDS service name change. I think in those cases, active/draining should both combine to form a unified cluster stats. I've filed a bug for this at https://github.com/envoyproxy/envoy/issues/12823.

Reply:
@htcuh Alright, that would affect at most how dropped requests will be reported. If Enovy choose to report stats via a unified cluster stats, then we need to change our existing behavior as well.

Back to our original question, does Envoy reset circuit breakers when receiving an CDS response that changes the EDS service name? This is more crucial for us about how the max-request enforcement should be applied.

Reply:
I'm pretty sure it does, since the circuit breaker resource tracking lives in the cluster, and EDS service name change implies a new cluster.

Reply:
Great, then that is clear enough to us for our existing issue. 

@markdroth You are right, we would need separate counters for each EDS LB to keep track of its in-flight RPCs.

Reply:
(Sorry for taking so long to get back to this thread -- I've been really swamped with other work for the last couple of weeks.)

If I understand all of this discussion correctly, here's where things stand:

Based on the description in envoyproxy/envoy#12823, it seems like the LRS behavior that Envoy wants is what gRPC already has, which is to actually use both the cluster name and EDS service name as the key for the LRS data.  When the EDS service name changes, any activity for the old EDS service (e.g., drops recorded before the EDS service name changed or long-running streams started before the change) will be reported under the old name, and any activity for the new EDS service name will be reported under the new name.  So we don't need any changes to this existing code in gRPC.

Since any change in the Cluster resource basically causes Envoy to create a new Cluster instance, warm it, and then swap it out for the old one, that does imply that there is a separate counter of active requests for each Cluster instance.  So if gRPC tracks these drops in the EDS policy, then each EDS policy instance can have its own counter, and the counter will implicitly reset whenever we swap out the EDS policy, which is what we want.  And as an added bonus, I think this is much simpler for us to implement than any other behavior.

The only behavior difference between gRPC and Envoy will be that if there is a CDS update that changes some field *other* than the EDS service name (e.g., if it changes the TLS certificate config), Envoy would reset this counter, whereas gRPC would not.  But that actually seems better -- in principle, we should not reset this counter if the thing that changed was not part of the key for reporting the data to LRS.  However, it is worth noting that this could result in possibly surprising behavior: if the current number of RPCs in flight is 105 and we receive a CDS update lowering the limit to 100, we will immediately start failing all RPCs until 6 of the currently in-flight RPCs are finished.  Envoy would actually reset the counter here, which means that even if none of the original 105 RPCs ever completed, it would allow 100 new RPCs to be started.  I think this is okay, but we should probably call this out in the doc.

@voidzcy I think we need the following changes:
- Please add a note about the behavior difference I described above.
- Please remove the paragraph starting on line 105.

I think that with those changes, the gRFC should reflect what we want here.  Thanks to everyone for the discussion!

Reply:
Thanks for looking into this.

> if the current number of RPCs in flight is 105 and we receive a CDS update lowering the limit to 100, we will immediately start failing all RPCs until 6 of the currently in-flight RPCs are finished. Envoy would actually reset the counter here, which means that even if none of the original 105 RPCs ever completed, it would allow 100 new RPCs to be started.

I think some concepts are distorted in this description. Since Enovy _replaces the cluster_ when receiving a lowered circuit breaker limit, logically there are _old_ and _new_ clusters. New RPCs are sent through the _new cluster_ immediately (as connections are pooled) and the _new cluster_ starts with new circuit breaker tracking states. Then definitely another 100 new RPCs are allowed. But in gRPC there is only one logical cluster during this whole process, so new RPCs just continue the existing cluster's circuit breaker tracking states. So the problem is not because of how circuit breakers are applied and implemented, but really is because of the difference in the concept of _logical clusters_.

Reply:
I added the description in terms of this behavior difference. I think we should focusing on describing the high-level design and concepts instead of trying to clarify every single piece of detail.

Reply:
The way you described this looks good.  I think it's primarily important to point out the actual behavior difference with this example, but also explaining the architectural reasons why the behavior difference occurs is also very useful.